datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemCheckBoxGroupViewHolderFactory.kt

```diff
@@ -54,11 +60,27 @@ internal object QuestionnaireItemCheckBoxGroupViewHolderFactory :
           prefixTextView.visibility = View.GONE
         }
         val (questionnaireItem, _) = questionnaireItemViewItem
+        val choiceOrientation =
+          questionnaireItem.choiceOrientation ?: ChoiceOrientationTypes.VERTICAL
         questionTextView.text = questionnaireItem.localizedTextSpanned
-        checkboxGroup.removeAllViews()
-        questionnaireItemViewItem.answerOption.forEach { answerOption ->
-          populateViewWithAnswerOption(answerOption)
+        checkboxGroup.removeViews(1, checkboxGroup.childCount - 1)
+        flow.referencedIds = IntArray(0)
+        when (choiceOrientation) {
+          ChoiceOrientationTypes.HORIZONTAL -> {
+            flow.setOrientation(Flow.HORIZONTAL)
+            flow.setWrapMode(Flow.WRAP_CHAIN)
+          }
+          ChoiceOrientationTypes.VERTICAL -> {
+            flow.setOrientation(Flow.VERTICAL)
+            flow.setWrapMode(Flow.WRAP_NONE)
+          }
         }
+        // By default, checkbox starts from index 1
+        questionnaireItemViewItem.answerOption.forEachIndexed { index, answerOption ->
+          populateViewWithAnswerOption(answerOption, choiceOrientation, index)
+        }
+        val refIds = IntArray(questionnaireItemViewItem.answerOption.size) { it + 1 }
+        flow.referencedIds = refIds
```

Comment:
I suggest we use the API `View.generateViewId()` to generate view id rather than doing the book keeping ourselves. this way we don't have to keep doing the +1 business which is probably bug-prone.

In the following code, for each answer option, we are generating a view id -> answer option pair, we pass that to the function, and then at the end we collect the view ids to give them to the flow layout as reference ids.

```suggestion
        questionnaireItemViewItem
          .answerOption
          .map { answerOption -> View.generateViewId() to answerOption }
          .onEach { populateViewWithAnswerOption(it.first, it.second, choiceOrientation) }
          .map { it.first }
          .let { flow.referencedIds = it.toIntArray() }
```

Reply:
Wow, yeah that is definitely better

Reply:
glad you liked it